pkg/odo/cli/service/create.go

```diff
@@ -447,56 +450,57 @@ func NewCmdServiceCreate(name, fullName string) *cobra.Command {
 	return serviceCreateCmd
 }
 
-// Parses group and version values from the alm-example
-func groupVersionALMExample(example map[string]interface{}) (group, version string) {
-	apiVersion := example["apiVersion"].(string)
-	// use SplitN so that if apiVersion field's value is something like
-	// etcd.coreos.com/v1/beta1 then group's value ends up being etcd.cores.com
-	// and version ends up being v1/beta1
-	gv := strings.SplitN(apiVersion, "/", 2)
-
-	group, version = gv[0], gv[1]
-	return
-}
-
-func resourceFromCSV(csv olmv1alpha1.ClusterServiceVersion, crdName string) (resource string) {
-	for _, crd := range csv.Spec.CustomResourceDefinitions.Owned {
-		if crd.Kind == crdName {
-			resource = strings.Split(crd.Name, ".")[0]
-			return
-		}
+// validateMetadataInCRD validates if the CRD has metadata.name field and returns an error
+func validateMetadataInCRD(crd map[string]interface{}, serviceName string) error {
+	metadata, ok := crd["metadata"].(map[string]interface{})
+	if !ok {
+		// this condition is satisfied if there's no metadata at all in the provided CRD
+		return fmt.Errorf("Couldn't find \"metadata\" in the yaml. Need metadata.name to start the service")
 	}
-	return
-}
 
-func getAlmExample(almExamples []map[string]interface{}, crd, operator string) (map[string]interface{}, error) {
-	for _, example := range almExamples {
-		if example["kind"].(string) == crd {
-			return example, nil
-		}
+	if _, ok := metadata["name"].(string); ok {
+		// found the metadata.name; no error
+		return nil
 	}
-	return nil, errors.Errorf("Could not find example yaml definition for %q service in %q operator's definition.\nPlease provide a file containing yaml specification to start the service from operator\n", crd, operator)
+	return fmt.Errorf("Couldn't find metadata.name in the yaml. Provide a name for the service")
 }
 
-func doesCRExist(kind string, csvs *olmv1alpha1.ClusterServiceVersionList) (olmv1alpha1.ClusterServiceVersion, error) {
-	for _, csv := range csvs.Items {
-		for _, operatorCR := range csv.Spec.CustomResourceDefinitions.Owned {
-			if kind == operatorCR.Kind {
-				return csv, nil
+// setServiceName modifies the CRD to contain user provided name on the CLI
+// instead of using the default one in almExample
+func setServiceName(crd interface{}, name string) (map[string]interface{}, error) {
```

Comment:
I think we need to have a struct called `DynamicCRD` which holds metdata, kind and other such fields with the original `crd` as a `map[string]interface{}`. 
So something like
```
type DynamicCRD struct {
Metadata Metadata
Kind string
RawCRD map[string]interface{}
}
```
and then this struct could have `getServiceNameFromCRD`,` setServiceName`, `CheckCRExists` etc

All the functions added so far seem to indicate that this refactor is gonna help us so I would suggest you please do the refactor as part of this PR 🙂 . And this change is related to most of the code added as well.

If you find any other opportunity to refactor then please do that too as part of this PR. And we shouldn't delay visible refactoring opportunities because else we would never do them.

Reply:
> So something like
> 
> ```
> type DynamicCRD struct {
> Metadata Metadata
> Kind string
> RawCRD map[string]interface{}
> }
> ```
> 
> and then this struct could have `getServiceNameFromCRD`,` setServiceName`, `CheckCRExists`(not sure about if this fits but consider if it does) etc

Makes sense to me. I'll make the change and ping for review after that.

> All the functions added so far seem to indicate that this refactor is gonna help us so I would suggest you please do the refactor as part of this PR :slightly_smiling_face: . And this change is related to most of the code added as well.
> 
> If you find any other opportunity to refactor then please do that too as part of this PR. And we shouldn't delay visible refactoring opportunities because else we would never do them.

We're on the same page here. https://github.com/openshift/odo/pull/3509/#pullrequestreview-444494596